File: js_basic_problems/22_valid_parentheses.js
// Problem 22: Valid Parentheses
// Input:
// {[()]}
// Output:
// YES

const fs = require("fs");
const data = fs.readFileSync(0, "utf-8").trim().split("\n");
let s = data[0].trim();
let stack = [];
let match = { '(': ')', '[': ']', '{': '}' };
for (let ch of s) {
    if (match[ch]) {
        stack.push(ch);
    } else {
        if (stack.length === 0 || match[stack.pop()] !== ch) {
            console.log("NO");
            return;
        }
    }
}
console.log(stack.length === 0 ? "YES" : "NO");

File: py_basic_problems/22_valid_parentheses.py
# Problem 22: Valid Parentheses
# Input:
# {[()]}
# Output:
# YES

s = input().strip()
stack = []
match = {'(': ')', '[': ']', '{': '}'}
for c in s:
    if c in match:
        stack.append(c)
    elif not stack or match[stack.pop()] != c:
        print("NO")
        break
else:
    print("YES" if not stack else "NO")
